exonum-java-binding-core/src/test/java/com/exonum/binding/storage/indices/ListSpliteratorTest.java

```diff
@@ -0,0 +1,303 @@
+    ListIndex<Integer> list = mock(ListIndex.class);
+    long size = 10;
+    long initialSize = size - 1;
+    when(list.size()).thenReturn(initialSize);
+    ModificationCounter counter = new IncrementalModificationCounter();
+
+    // Create a spliterator
+    Spliterator<Integer> spliterator = new ListSpliterator<>(list, counter, true);
+
+    // "Add" an element to the list
+    when(list.size()).thenReturn(size);
+    counter.notifyModified();
+
+    // Test that spliterator is usable despite modifications after initialization but before
+    // binding, reporting the correct size
+    assertThat(spliterator.estimateSize()).isEqualTo(size);
+  }
+
+  private static void assertHasDetectedModification(Spliterator<Integer> spliterator) {
+    assertThrows(ConcurrentModificationException.class,
+        () -> spliterator.tryAdvance(NULL_CONSUMER));
+    assertThrows(ConcurrentModificationException.class,
+        () -> spliterator.forEachRemaining(NULL_CONSUMER));
+    assertThrows(ConcurrentModificationException.class, spliterator::trySplit);
+  }
+
+  @ParameterizedTest
+  @ValueSource(longs = {0, 1, 2})
+  void estimateSizeAtBindTime(long listSize) {
+    ListIndex<Integer> list = mock(ListIndex.class);
+    when(list.size()).thenReturn(listSize);
+    ModificationCounter counter = mock(ModificationCounter.class);
+    Spliterator<Integer> spliterator = new ListSpliterator<>(list, counter, true);
+
+    assertThat(spliterator.estimateSize()).isEqualTo(listSize);
+  }
+
+  @ParameterizedTest
+  @ValueSource(longs = {1, 2, 3})
+  void estimateSizeAfterSingleSuccessfulAdvance(long listSize) {
+    ListIndex<Integer> list = mock(ListIndex.class);
+    when(list.size()).thenReturn(listSize);
+    ModificationCounter counter = mock(ModificationCounter.class);
+    Spliterator<Integer> spliterator = new ListSpliterator<>(list, counter, true);
+
+    // Advance the iterator
+    spliterator.tryAdvance(NULL_CONSUMER);
+
+    assertThat(spliterator.estimateSize()).isEqualTo(listSize - 1);
+  }
+
+  @Test
+  void estimateSizeIsZeroAfterSpliteratorIsConsumed() {
+    ListIndex<Integer> list = mock(ListIndex.class);
+    long listSize = 5;
+    when(list.size()).thenReturn(listSize);
+    ModificationCounter counter = mock(ModificationCounter.class);
+    Spliterator<Integer> spliterator = new ListSpliterator<>(list, counter, true);
+
+    // Advance the iterator
+    spliterator.forEachRemaining(NULL_CONSUMER);
+
+    assertThat(spliterator.estimateSize()).isEqualTo(0);
+  }
+
+  private static Spliterator<Integer> createSpliteratorOf(int[] source) {
+    ListIndex<Integer> list = mock(ListIndex.class);
+    lenient().when(list.get(anyLong())).thenAnswer((Answer<Integer>) invocation -> {
+      Long index = invocation.getArgument(0);
+      return source[Math.toIntExact(index)];
+    });
+    lenient().when(list.size()).thenReturn((long) source.length);
+
+    ModificationCounter modCounter = mock(ModificationCounter.class);
```

Comment:
Use `ImmutableModificationCounter` instead of mock here?

Reply:
ImmutableModificationCounter is package-private at the moment, I'd leave it as is :-)